doc/cascadia/Keybindings-Arguments.md

```diff
@@ -0,0 +1,291 @@
+runtimeclass NewTabWithProfileEventArgs : IKeyBindingArgs
+{
+    Int32 ProfileIndex;
+}
+
+[default_interface]
+runtimeclass AppKeyBindings : Microsoft.Terminal.Settings.IKeyBindings
+{
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, CopyTextEventArgs> CopyText;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabEventArgs> NewTab;
+    event Windows.Foundation.TypedEventHandler<AppKeyBindings, NewTabWithProfileEventArgs> NewTabWithProfile;
+```
+
+We'll also change our existing map in the `AppKeyBindings` implementation.
+Currently, it's a `std::unordered_map<KeyChord, ShortcutAction, ...>`, which
+uses the `KeyChord` to lookup the `ShortcutAction`. We'll need to introduce a
+new type `ActionAndArgs`:
+
+```C#
+runtimeclass ActionAndArgs
+{
+    ShortcutAction action;
+    IKeyBindingArgs args;
+}
+```
+
+and we'll change the map in `AppKeyBindings` to a `std::unordered_map<KeyChord,
+ActionAndArgs, ...>`.
+
+When we're parsing keybindings, we'll need to construct args for each of the
+events to go with each binding. When we find some key chord bound to a given
+Action, we'll construct the `IKeyBindingArgs` for that action. FOr many actions,
+these args will be an empty class. However, when we do find an action that needs
+additional parsing, `AppKeyBindingsSerialization` will do the extra work to
+parse the args for that action.
+
+Once the `IKeyBindingArgs` is built for the keybinding, we'll set it in
+`AppKeyBindings` with a updated `AppKeyBindings::SetKeyBinding` call.
+`SetKeyBinding`'s signature will be updated to take a `ActionAndArgs` instead.
+
+### Executing KeyBinding Actions with Arguments
+
+When we're handling a keybinding in `AppKeyBindings::_DoAction`, we'll trigger
+the event handlers with the `IKeyBindingArgs` we've stored in the map with the
+`ShortcutAction`.
+
+Then, in `App`, we'll handle each of these events. We set up lambdas as event
+handlers for each event in `App::_HookupKeyBindings`. In each of those
+functions, We'll inspect the `IKeyBindingArgs` parameter, and use args from its
+implementation to call callbacks in the `App` class.
+
+Instead of:
+```c++
+    void App::_HookupKeyBindings(TerminalApp::AppKeyBindings bindings) noexcept
+    {
+        // ...
+        bindings.NewTabWithProfile([this](const auto index) { _OpenNewTab({ index }); });
+    }
+```
+
+The code will look like:
+
+```c++
+    void App::_HookupKeyBindings(TerminalApp::AppKeyBindings bindings) noexcept
+    {
+        // ...
+        bindings.NewTabWithProfile([this](const auto& sender, const auto& args) { _OpenNewTab({ args.ProfileIndex() }); });
+    }
+```
+
+We could even go so far as to update `App` to have methods defined with the
+actual keybinding function signatures, so instead we'd just have:
+
+
+```c++
+    void App::_HookupKeyBindings(TerminalApp::AppKeyBindings bindings) noexcept
+    {
+        // ...
+        bindings.NewTabWithProfile({ this, &App::_OpenNewTab });

```

Comment:
I much prefer this to the above.

Reply:
The above one, the lambda-ness of it is getting insane. There are so many ({{({()})}}}) that I think it will lead to mistakes and issues more than separating out the method. You should prescribe a separate method if possible at this point.

Reply:
Yea as I was writing it I realized this was possible, which is just infinitely better.